File: video_stab.py
import cv2
import numpy as np

CAM = 0

class Tracker:
    def __init__(self):
        self.tracked_features = []
        self.prev_gray = None
        self.fresh_start = True
        # 3x3 identity matrix for rigid transform (affine 2x3 in a 3x3 matrix)
        self.rigid_transform = np.eye(3, dtype=np.float32)

    def process_image(self, img):
        # Convert to grayscale
        gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
        
        # If we need more features
        if len(self.tracked_features) < 200:
            # goodFeaturesToTrack returns array of shape (N, 1, 2)
            corners = cv2.goodFeaturesToTrack(gray, maxCorners=300, qualityLevel=0.01, minDistance=1)
            if corners is not None:
                print(f"found {len(corners)} features")
                # Convert corners to list of (x,y) points
                corners = corners.reshape(-1, 2)
                self.tracked_features.extend(corners.tolist())

        if self.prev_gray is not None and len(self.tracked_features) > 0:
            # Convert tracked_features to numpy array for optical flow
            prev_pts = np.array(self.tracked_features, dtype=np.float32).reshape(-1, 1, 2)
            
            # Calculate optical flow
            corners, status, errors = cv2.calcOpticalFlowPyrLK(
                self.prev_gray, gray, prev_pts, None, winSize=(10, 10)
            )
            
            # Count successful tracks
            successful_tracks = np.count_nonzero(status)
            
            if successful_tracks < len(status) * 0.8:
                print("cataclysmic error")
                self.rigid_transform = np.eye(3, dtype=np.float32)
                self.tracked_features.clear()
                self.prev_gray = None
                self.fresh_start = True
                return
            else:
                self.fresh_start = False

            # Estimate rigid transform
            good_prev = prev_pts[status == 1].reshape(-1, 2)
            good_new = corners[status == 1].reshape(-1, 2)
            
            new_rigid_transform = cv2.estimateAffine2D(good_prev, good_new)[0]
            if new_rigid_transform is not None:
                # Convert 2x3 transform to 3x3
                nrt33 = np.eye(3, dtype=np.float32)
                nrt33[0:2, :] = new_rigid_transform
                self.rigid_transform = self.rigid_transform @ nrt33

            # Update tracked features
            self.tracked_features = good_new.tolist()

        # Draw tracked features
        for point in self.tracked_features:
            x, y = map(int, point)
            cv2.circle(img, (x, y), 3, (0, 0, 255), -1)  # -1 for filled circle

        # Update previous frame
        self.prev_gray = gray.copy()

def main():
    vc = cv2.VideoCapture(CAM)
    tracker = Tracker()
    
    print("in main")

    while True:
        ret, frame = vc.read()
        if not ret:
            break
            
        orig = frame.copy()
        tracker.process_image(orig)

        # Calculate inverse transform and warp
        inv_trans = cv2.invertAffineTransform(tracker.rigid_transform[0:2, :])
        orig_warped = cv2.warpAffine(orig, inv_trans, (frame.shape[1], frame.shape[0]))

        cv2.imshow("orig", orig_warped)
        if cv2.waitKey(1) & 0xFF == 27:  # Exit on ESC
            break

    vc.release()
    cv2.destroyAllWindows()

if __name__ == "__main__":
    main()

File: video_stab.cpp
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/features2d/features2d.hpp>
#include <opencv2/video/video.hpp>
#include <iostream>

#define CAM 0

using namespace cv;
using namespace std;

class Tracker {
    vector<Point2f> trackedFeatures;
    Mat             prevGray;
public:
    bool            freshStart;
    Mat_<float>     rigidTransform;

    Tracker():freshStart(true)
    {
        rigidTransform = Mat::eye(3,3,CV_32FC1); //affine 2x3 in a 3x3 matrix
    }

    void processImage(Mat& img)
    {
        Mat gray; cvtColor(img,gray,CV_BGR2GRAY);
        vector<Point2f> corners;
        if(trackedFeatures.size() < 200)
        {
            goodFeaturesToTrack(gray,corners,300,0.01,1);
            cout << "found " << corners.size() << " features\n";
            for (unsigned int i = 0; i < corners.size(); ++i)
            {
                trackedFeatures.push_back(corners[i]);
            }
        }

        if(!prevGray.empty()) {
            vector<uchar> status; vector<float> errors;
            calcOpticalFlowPyrLK(prevGray,gray,trackedFeatures,corners,status,errors,Size(10,10));

            if(countNonZero(status) < status.size() * 0.8)
            {
                cout << "cataclysmic error \n";
                rigidTransform = Mat::eye(3,3,CV_32FC1);
                trackedFeatures.clear();
                prevGray.release();
                freshStart = true;
                return;
            }
            else
                freshStart = false;

            Mat_<float> newRigidTransform = estimateRigidTransform(trackedFeatures,corners,false);
            Mat_<float> nrt33 = Mat_<float>::eye(3,3);
            newRigidTransform.copyTo(nrt33.rowRange(0,2));
            rigidTransform *= nrt33;

            trackedFeatures.clear();
            for (unsigned int i = 0; i < status.size(); ++i)
            {
                if(status[i]) {
                    trackedFeatures.push_back(corners[i]);
                }
            }
        }

        for (unsigned int i = 0; i < trackedFeatures.size(); ++i)
        {
            circle(img,trackedFeatures[i],3,Scalar(0,0,255),CV_FILLED);
        }

        gray.copyTo(prevGray);
    }
};


int main()
{
    VideoCapture vc(CAM);

    Mat frame,orig,orig_warped,tmp;

    Tracker tracker;

    cout << "in main" << endl;

    while(1)
    {
        vc >> frame;
        if(frame.empty()) break;
        frame.copyTo(orig);

        tracker.processImage(orig);

        Mat invTrans = tracker.rigidTransform.inv(DECOMP_SVD);
        warpAffine(orig,orig_warped,invTrans.rowRange(0,2),Size());

        imshow("orig",orig_warped);
        waitKey(1);
    }
    vc.release();
}
